GPG-643 Add script to deploy IP denylist app

AlexJacksonDS opened the pull request:
Initial state of it, not ready to merge.

AlexJacksonDS commented:
Updated to add the IPs as an input parameter as I couldn't get the config vcap way to work. Added a readme which hopefully explains how to use it sufficiently. Should all be fine now

Binney commented on Infrastructure/gpg-ipdeny/manifest.yml:
```diff
@@ -0,0 +1,15 @@
+applications:
+  - name: gpg-ipdeny
+    instances: 1
+    memory: 256M
```
We will want to think about this number (and this app generally) during the load test - as all our traffic will be going through it. Not one for this review but for a few weeks' time

Binney commented on Infrastructure/gpg-ipdeny/nginx.conf:
```diff
@@ -0,0 +1,107 @@
+                         ' "requestTime": $request_time, '
+                         ' "httpHost": "$http_host"}';
+
+  access_log /dev/stdout access_json;
```
Looks like these will just sit around on the instance and we don't send them anywhere? While I don't think it adds a lot to send the whole thing to logit, I would be very interested in stats on how many users we are denying access to - is that something we can pipe to another log stream and consider uploading to the cloud along with all the app server logs?

Binney commented on Infrastructure/gpg-ipdeny/nginx.conf:
```diff
@@ -0,0 +1,107 @@
+      proxy_http_version 1.1;
+      proxy_ssl_server_name on;
+      proxy_ssl_protocols TLSv1.2;
+      proxy_set_header Connection "";
```
Any other headers we're expecting to need to set here? e.g. application login token? (Is there a way to make it pass everything/is that desirable?)

Binney commented on Infrastructure/gpg-ipdeny/deploy.sh:
```diff
@@ -0,0 +1,79 @@
+  exit 1
+fi
+
+while getopts ":a:e:i:r:s:" opt; do
```
Niceeeeeeeee

Binney commented on Infrastructure/gpg-ipdeny/README.md:
````diff
@@ -0,0 +1,44 @@
+./deploy.sh -e gpg-dev -r gpg-dev-ipdeny -s gpg-dev-ipdeny-service -a gender-pay-gap-dev -i "1.2.3.4,2.3.4.5"
+```
+
+The comma separated list of banned IPs can be found in Zoho, when updating this list, newlines should be avoided as they can cause syntax errors within the nginx.conf. This can be seen by an error when deploying, and then if you try and run
````
The list in Zoho is quite long and I can see it being quite hard to maintain. With that in mind I wonder how much extra work it would be to use a file here - e.g.
* Argument is a filename
* File is basically "exactly what you found in Zoho" (i.e. what's in Zoho would need editing to remove comments)
* File comprises one entry per line, meaning it's easier to read/maintain
* Script replaces newlines with commas

thoughts?

AlexJacksonDS replied:
Agreed that it would be better to do something like that, would just need to figure out how

Binney commented on Infrastructure/gpg-ipdeny/README.md:
```diff
@@ -0,0 +1,44 @@
+# gpg-ipdeny
```
So the other place that could use some documentation is... there's a script at `Infrastructure/CreateEnvironment.sh` (and corresponding `DeleteEnvironment.sh` which could use updating in line with this. @dulciejackson will be putting them to the test when she builds the load test environment, so they don't necessarily need to be tested but at least a rough guess at the correct commands - which will error if it doesn't work, so we don't forget to fix it.

Binney reviewed (approved):
Lovely stuff :+1: Main thoughts are the maintainability of the denylist file, and how we go about testing this in preprod. In particular you should liaise with Dulcie around what this extra infrastructure means for the load test. Looking forward to getting it up and running!

AlexJacksonDS commented:
Updated to take a filename parameter instead of a comma separated list of IP addresses (also updated the file in zoho to match)
Updated the nginx.conf so that headers are preserved in responses rather than being changed (particularly cache-control)

Binney commented on Infrastructure/gpg-ipdeny/nginx.conf:
```diff
@@ -90,7 +90,7 @@ http {
         set $cf_forwarded_uri /$3;
       }
-
+ 
```
Some weird whitespace here

Binney commented on Infrastructure/gpg-ipdeny/nginx.conf:
```diff
@@ -101,6 +101,12 @@ http {
+      # Setting this allows the asp.net app to set the cache-control response headers
+      expires off;
+
+      # Removes the Server: nginx header from the response
```
Niceeeee

Binney commented on Infrastructure/gpg-ipdeny/README.md:
````diff
@@ -20,25 +20,16 @@ Before running the deployment script you must be logged into GOV.UK PaaS using t
+For example to set up ip denylisting on the dev space, if you've created/placed the denylist file in the same directory as the deployment script, the command could be run (using the gitbash window set up with jq earlier) as 
 ```
-./deploy.sh -e gpg-dev -r gpg-dev-ipdeny -s gpg-dev-ipdeny-service -a gender-pay-gap-dev -i "1.2.3.4,2.3.4.5"
+./deploy.sh -e gpg-dev -r gpg-dev-ipdeny -s gpg-dev-ipdeny-service -a gender-pay-gap-dev -f "GPG IP Denylist.txt"
````
I'm not a big fan of spaces in filenames but not massively fussed about changing it

Binney commented on Infrastructure/gpg-ipdeny/deploy.sh:
```diff
@@ -43,16 +43,16 @@ if [ -z "${PROTECTED_APP_SPACE_NAME+set}" ]; then
 NGINX_DENY_STATEMENTS=""
 for addr in "${IPS[@]}";
-  do NGINX_DENY_STATEMENTS="$NGINX_DENY_STATEMENTS deny $addr;"; true;
+  do NGINX_DENY_STATEMENTS="$NGINX_DENY_STATEMENTS deny ${addr//[$'\r']};"; true;
```
Oh yeah I like this way more than commas

Binney replied:
Is it worth windows-proofing? will you get any line ending issues?

AlexJacksonDS replied:
I did have some issues with the line ending hence the ${addr//[$'\r']}, I think the readarray was just splitting on and stripping the \n of the \r\n in the text file. If we find we do need to strip other stuff it should just be a case of adding them here

Binney replied:
That works for me, as long as there isn't something easy we should be doing to make it a bit more resilient

Binney reviewed (approved):
Aceeee big fan :+1: 

How much testing have you been able to do with the headers - is there anything else we think it is overwriting/blocking? Or is it just the cache and server headers?

AlexJacksonDS commented:
> How much testing have you been able to do with the headers - is there anything else we think it is overwriting/blocking? Or is it just the cache and server headers?

Headers all look the same on dev and preprod now (a few slightly different values for things like content-length) and everything seems to be working fine, e.g. it hasn't re-broken the secure cookies by unsetting the x-forwarded-proto headers. I haven't checked every page though

dulciejackson reviewed (approved):
Looks great 😄 
I'll chat with you about anything we'll need to configure with this once I get round to running a load test!
